katpoint/projection.py

```diff
@@ -129,24 +133,196 @@
+    return x, y
+
+
+def sphere_to_ortho(az0, el0, az, el, min_cos_theta=None):
+    """Do calculations common to all zenithal/azimuthal projections.
+
+    This does a basic orthographic (SIN) projection and also returns the angular
+    separation / native latitude `theta` in cosine form, which can be used to
+    construct many other projections. The angular separation is optionally
+    checked against a projection-specific limit, and the (x, y) outputs are
+    treated accordingly.
+
+    Parameters
+    ----------
+    az0 : float or array
+        Azimuth / right ascension / longitude of reference point(s), in radians
+    el0 : float or array
+        Elevation / declination / latitude of reference point(s), in radians
+    az : float or array
+        Azimuth / right ascension / longitude of target point(s), in radians
+    el : float or array
+        Elevation / declination / latitude of target point(s), in radians
+    min_cos_theta : float, optional
+        Limit on angular separation of target and reference, used to clip (x, y)
+
+    Returns
+    -------
+    ortho_x : float or array
+        Azimuth-like coordinate(s) on plane (equivalent to l), in radians
+    ortho_y : float or array
+        Elevation-like coordinate(s) on plane (equivalent to m), in radians
+    cos_theta : float or array
+        Angular separation of target and reference points, expressed as cosine
+    """
+    # Ensure that elevation angles are in valid range if they are finite numbers
+    check = 'Elevation angle outside range of +- pi/2 radians'
+    el0 = OutOfRange.treat(el0, check, lower=-np.pi / 2.0, upper=np.pi / 2.0)
+    el = OutOfRange.treat(el, check, lower=-np.pi / 2.0, upper=np.pi / 2.0)
     sin_el, cos_el, sin_el0, cos_el0 = np.sin(el), np.cos(el), np.sin(el0), np.cos(el0)
-    # Keep azimuth delta between -pi and pi - probably unnecessary, but the only normalisation of az inputs
+    # Keep azimuth delta between -pi and pi - probably unnecessary,
+    # but the only normalisation of az inputs
     delta_az = (az - az0 + np.pi) % (2.0 * np.pi) - np.pi
     sin_daz, cos_daz = np.sin(delta_az), np.cos(delta_az)
     # Theta is the native latitude (0 at reference point, increases radially outwards)
     cos_theta = sin_el * sin_el0 + cos_el * cos_el0 * cos_daz
-    # Do basic orthographic projection: x = sin(theta) * sin(phi), y = sin(theta) * cos(phi)
+    # Do basic orthographic projection:
+    # x = sin(theta) * sin(phi), y = sin(theta) * cos(phi)
     ortho_x = cos_el * sin_daz
     ortho_y = sin_el * cos_el0 - cos_el * sin_el0 * cos_daz
+    if min_cos_theta is not None:
+        check = ('Target point more than {} pi radians away from '
+                 'reference point'.format(np.arccos(min_cos_theta) / np.pi))
+        cos_theta = OutOfRange.treat(cos_theta, check, lower=min_cos_theta)
+        sin_theta = np.sqrt(1.0 - cos_theta * cos_theta)
+        ortho_x, ortho_y = safe_scale(ortho_x, ortho_y, new_radius=sin_theta)
```

Comment:
No, `cos_theta` could be substantially adjusted by the out-of-range check (from e.g. -1 to 0), in which case `sin_theta` would not match the current length of `(ortho_x, ortho_y)`. The scaling rectifies this so that `(x, y, cos_theta)` is internally consistent and valid (actually a *3D unit vector*, now that I think about it - missed that connection!)

This is really the key of the `clip` method (and the `nan` method, for that matter). Previously I only checked `cos_theta` but did not modify `(x, y)` accordingly.

I could call `cos_theta` `new_cos_theta` or `treated_cos_theta` at the end, if it clarifies matters.

Reply:
Added a comment to clarify that this is a new `sin_theta` due to clipping or NaNs.